Support overriding `speaker_mode` in `AudioServer`

BuzzLord opened the pull request:
**Updated**: _Refactored this to only handle setting/overriding the `speaker_mode` of a project. Channel remapping will be moved to another PR._

Adds support (via project settings under `audio/general/speaker_mode_override`) to the `AudioServer` to change the speaker mode (i.e. number of channels) that Godot uses internally, independent of the audio driver used. It can also be set by command line argument `--speaker-mode-override`, which will override the project setting.

It can be used to test more channels (e.g. 7.1 surround when using a stereo audio setup), or fewer channels (e.g. stereo when using a 7.1 surround setup). This is primarily intended to be used to help troubleshoot audio issues when testing with different speaker setups, or hard coding a speaker mode for a project. But it can also be used by end users to tweak how the engine behaves for their specific hardware, using the command line arg.

~~This PR also adds the ability to remap internal channels onto different output channels of the audio device. For example, you can remap the rear or side stereo channels of a 7.1 surround system into the front channels of a stereo system in order to verify the audio is correct.~~

This is a extension of, and replacement for #105682. Like that PR, this feature makes it possible for Godot developers without the necessary surround sound peripheral to reproduce bugs and review fixes to issues that are the result of an unanticipated number of channels, such as: #92532

BuzzLord committed:
Add support for overriding speaker_mode.

Add support (via project settings) to AudioServer to change the speaker_mode
that Godot uses internally, independent of the audio driver used. Can accept a 'Default'
value, using the audio driver speaker_mode, or use the specified mode.

This is intended to be used to help troubleshoot audio issues when testing with different
speaker setups, but can be used to force a specific speaker_mode on projects where
appropriate.

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

BuzzLord commented:
Updated based on the feedback from the most recent audio team meeting. I left the new commit as is for now (i.e. didn't `--amend`), to make it clear what I changed. I'll rebase and squash it into one commit once its done.

I also made a little project to help show off the functionality.

[test-godot-channel-override.zip](https://github.com/user-attachments/files/20245111/test-godot-channel-override.zip)

adamscott commented:
> > * the other one for overriding speaker channels (which needs tweaking, as it cannot really reside in the project configs, as discussed in the last audio meeting)
> 
> It has been with the most recent update. Overriding is not in project settings, its done with get/set methods on AudioServer.

Ohhh, I see. The settings I saw in the project were lingering options then. Sorry.

BuzzLord commented:
Refactored this to only handle setting/overriding the `speaker_mode` of a project. 

Channel remapping will be moved to another PR.

adamscott reviewed (dismissed):
Discussed during the 2025-06-24 audio meeting. Seems ok so far. Would be great to add it to 4.5. This is a feature after the feature freeze, but it's quite a small change for the functionality it brings to the table.
Will test it and comment afterwards later today.

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -464,6 +464,10 @@
 			Sets the [url=https://developer.apple.com/documentation/avfaudio/avaudiosessioncategory]AVAudioSessionCategory[/url] on iOS. Use the [code]Playback[/code] category to get sound output, even if the phone is in silent mode.
 		</member>
+		<member name="audio/general/speaker_mode" type="int" setter="" getter="" default="-1">
+			The speaker mode used internally by Godot. [code]Default[/code] will use the audio drivers speaker mode, while other values will override the internal speaker mode, and thus the number of audio channels, used by the [AudioServer].
```
```suggestion
			The speaker mode used internally by Godot. [code]Default[/code] will use the audio driver's speaker mode, while other values will override the internal speaker mode, and thus the number of audio channels, used by the [AudioServer].
```

AThousandShips commented on doc/classes/ProjectSettings.xml:
```diff
@@ -464,6 +464,10 @@
 		</member>
+		<member name="audio/general/speaker_mode" type="int" setter="" getter="" default="-1">
+			The speaker mode used internally by Godot. [code]Default[/code] will use the audio drivers speaker mode, while other values will override the internal speaker mode, and thus the number of audio channels, used by the [AudioServer].
+			When the internal speaker mode differs from the drivers speaker mode, audio channels will be muted, ignored, or mapped to the audio drivers channels.
```
```suggestion
			When the internal speaker mode differs from the driver's speaker mode, audio channels will be muted, ignored, or mapped to the audio driver's channels.
```

AThousandShips commented on servers/audio_server.cpp:
```diff
@@ -264,7 +264,7 @@ void AudioServer::_driver_process(int p_frames, int32_t *p_buffer) {
 
 	if (channel_count != get_channel_count()) {
-		// Amount of channels changed due to a output_device change
+		// Amount of channels changed due to a output_device change, or speaker_mode change
```
```suggestion
		// Number of channels changed due to a output_device or speaker_mode change
```

BuzzLord commented:
After today's discussion, it seems that we're leaning towards the idea that this PR probably only applies to developers troubleshooting engine issues, and not as a useful feature to be exposed in games. It was suggested that this could be done from the OS level, so I did some poking around to try it.

I was able to use the [VB-Audio VB-CABLE driver](https://vb-audio.com/Cable/index.htm) to effectively do what this PR does. It should work for Windows and Mac. Linux almost certainly also has an equivalent feature.

For Windows:
1. Install the VB-CABLE driver.
2. In the **Sound** configuration window, under the _CABLE Output_ Recording device properties, enable _Listen to this device_, and set the playback device to the actual output device.
3. Configure _CABLE Input_ to be an appropriate speaker configuration (e.g. 7.1 Surround).
4. Set _CABLE Input_ to be the default playback device.

This results in the engine using 7.1 surround, and I confirmed the `AudioEffectCapture`/`AudioEffectRecord` bugs can be reproduced this way on a stereo audio device.

Would it make more sense to just write some documentation describing how to configure these external methods rather than include this PR in the engine itself?

fdemir commented:
Bump! We need this!!!!

BuzzLord commented:
> Bump! We need this!!!!

Do you mean the PR, or the proposed documentation in my last comment? We decided against merging the PR, due to its limited use as a debug tool, primarily for engine devs. If it is the PR you want, what exactly is it that you're hoping it can help with?

If its the docs you're waiting for, I've just been too busy/distracted recently to write anything detailed, but the instructions above are the core of what I envision, just with extra screenshots of the process.

BuzzLord commented:
Closing this PR as it's not going to be merged. I (or someone) can make a new PR for the proposed documentation describing how to setup and test different audio configurations (at least for the main dev platforms Win/Linux/Mac).

BuzzLord committed:
Added command line parameter.

The speaker_mode can now be overridden from the command line.
Changed how the numeric value is used so 0 is default.

BuzzLord committed:
Changed setting name to 'speaker-mode-override'.

BuzzLord committed:
Moved speaker_mode_override to audio/driver project settings.

While not directly changing anything on the driver, it does interact with
the driver, and mentions it in the docs.

BuzzLord committed:
Added shell completion docs.

Ivorforce commented on doc/classes/ProjectSettings.xml:
```diff
@@ -461,6 +461,11 @@
+		<member name="audio/driver/speaker_mode_override" type="int" setter="" getter="" default="0">
+			Override the speaker mode used internally by Godot. [code]Disabled[/code] will use the audio driver's speaker mode, while other values will override the internal speaker mode, and thus the number of audio channels, used by the [AudioServer].
+			When the [AudioServer] speaker mode is overridden and differs from the audio driver's speaker mode, audio channels will be ignored or muted, depending on if there are more channels than the driver supports, or fewer (respectively).
+			This setting can itself be overridden using the [code]--speaker-mode-override &lt;mode&gt;[/code] command line argument.
```
I think this needs a brief description of why someone would use this.

BuzzLord replied:
I added another sentences here:
`Override the speaker mode used internally by Godot. Can be used to enforce a particular speaker mode (e.g. stereo in a 2D game), or to help test or troubleshoot different speaker configurations.`

Ivorforce reviewed (commented):
Looks good overall! I've got mainly a few nitpicks.

BuzzLord committed:
Minor tweaks based on PR feedback.

BuzzLord committed:
Shifted numeric override values down one.

Numeric value of internal speaker_mode_override now matches the
AudioServer SpeakerMode enum. Changed CLI interface to use string
values (more closely matching the behavior of other enum CLI
options), so now the numeric value is hidden from users (which was
my the reason for the shift up in the first place).
